clang/test/SemaCXX/unary-expr-or-type-trait-invalid.cpp

```diff
@@ -0,0 +1,26 @@
+// RUN: %clang_cc1 -fsyntax-only -verify %s
+
+a() {struct b c (sizeof(b * [({ {tree->d* next)} 0
+
+// expected-error@3 {{a type specifier is required for all declarations}}
+// expected-error@3 {{use of undeclared identifier 'tree'; did you mean 'true'?}}
+// expected-error@3 {{member reference type 'bool' is not a pointer}}
+// expected-error@3 {{expected ';' after expression}}
+// expected-error@3 {{use of undeclared identifier 'next'; did you mean 'new'?}}
+// expected-error@3 {{expected expression}}
+// expected-error@3 {{expected ';' after expression}}
+// expected-error@26 {{expected '}'}}
+// expected-note@3 {{to match this '{'}}
+// expected-error@26 {{expected ')'}}
+// expected-note@3 {{to match this '('}}
+// expected-error@26 {{expected ']'}}
+// expected-note@3 {{to match this '['}}
+// expected-error@26 {{expected ')'}}
+// expected-note@3 {{to match this '('}}
+// expected-error@3 {{using declaration 'exp' instantiates to an empty pack}}
+// expected-error@3 {{variable has incomplete type 'struct b'}}
+// expected-note@3 {{forward declaration of 'b'}}
+// expected-error@3 {{expected ';' at end of declaration}}
+// expected-error@26 {{expected '}'}}
+// expected-note@3 {{to match this '{'}}
+// expected-warning@3 {{expression result unused}}
```

Comment:
CC @Endilll @erichkeane  Do we have a cleaner way of writing fuzzer tests? Because this is rather atrocious to look at.

Reply:
OOof, that is awful.  Is that really the minimal example that causes the crash?  I don't really have a great idea.  I might disable the warning with a `-Wno` flag, but most of those are not actually important to what we want to do here.  The whole point is to ensure it doesn't crash/assert, not to make sure we emit all those diagnostics.  

We should come up with a way to test "did not crash" instead of this.

Reply:
Instead of running `-verify` on the test, can we just run `FileCheck` and ensure there are no strings like `PLEASE submit a bug report` or something?

Reply:
Not really, not all of the test configurations use the normal crash handler... I would hope the author could use the test as an example to come up with a minimal example that does the bare minimum to cause this issue rather than using the fuzz itself: I don't think the fuzz generated code should actually be in our test suite, just the actual use case that is required to get into this situation.

Reply:
Definitely agree that the test file is pretty ugly. I tried to remove some parts of the original code but none of my reductions kept the crash, so that's why I left it in full. I think ideally there would just be a way to make sure that there is no crash without checking the specific diagnostics, but it seems that option isn't available.

Reply:
The last time we discussed this, our testing infrastructure wasn't ready for regression tests that ensure we don't crash. Ideally we need to extend lit to check for a specific return code of a RUN line. The other way is to overspecify the test via `-verify`, and you can do stuff like `// expected-error-re@3 * {{}}` to make it somewhat acceptable.

Reply:
Thanks @Endilll for the suggestion. I think the cleanest approach right now is to just check for diagnostics without scrutinizing the specifics and I left a note to indicate the exact goal of the test.

Reply:
I think we need as a project to decide the best way forward for these sorts of tests.  I believe there is an RFC going around on fuzzed inputs, so I'm not really willing to accept this test as long as we haven't decided as a project how we wnat to handle this, particularly for one that is as ugly/overly-diagnosy as this one.

Reply:
One idea I had is we could add another mode to `-verify`: something like `// expected-no-diagnostics`, but instead it’d be `// expected-diagnostics` or something like that, and it passes if there is at least one diagnostic (and fails if we crash or if there is no diagnostic, of course).

Reply:
> One idea I had is we could add another mode to `-verify`: something like `// expected-no-diagnostics`, but instead it’d be `// expected-diagnostics` or something like that, and it passes if there is at least one diagnostic (and fails if we crash or if there is no diagnostic, of course).

In that case, we already have the ability to do something like: `// expected-error-re@* 0+ {{*}} //expected-warning-re@* 0+{{*}} // expected-note-re@* 0+{{*}}` 

Which does a regex match for 'everything', at line 'anything', 0 or more times.

Reply:
See #78898 for the previous discussion on the matter of tests based on crash reports.

Reply:
In most scenarios, we really do want to use `-verify` to see what the diagnostic behavior is because we want it to be very obvious when diagnostic behavior needs to be improved. (Similar for `-emit-llvm` when doing codegen tests.) However, we only care about the behavior of code that a real user is likely to write, which is not typical for fuzzer-generated crashing issues. I think it would be useful to have a special `-verify` mode that ignores all diagnostics (whether any are generated or not) and only tests that there was not a crash. Something like `// expected-no-crash` and we diagnose if any other `// expected-whatever` comments appear anywhere in the file. That should make it more clear "we're not interested in behavior beyond not crashing" for anyone reading the test.

Reply:
`// expected-no-crash` would be a good name for the directive, but I'm not sure that frontend is a good place for such check layering-wise.

Reply:
Yeah, it seems like more of a `lit` layer thing than a frontend layer thing.

So long as it's obvious that the intent of the test is "test this doesn't crash", I'm not too hung up on the syntax.

Reply:
At this point it seems like this is a separate issue, maybe we should open an issue for this and we can move the discussion there?